crates/accelerate/src/commutation_checker.rs

```diff
@@ -0,0 +1,770 @@
+            py,
+            first_op,
+            first_params,
+            first_qargs,
+            second_op,
+            second_params,
+            second_qargs,
+        )?;
+
+        // TODO: implement a LRU cache for this
+        if self.current_cache_entries >= self.cache_max_entries {
+            self.clear_cache();
+        }
+        // Cache results from is_commuting
+        let key1 = hashable_params(first_params)?;
+        let key2 = hashable_params(second_params)?;
+        self.cache
+            .entry((first_op.name().to_string(), second_op.name().to_string()))
+            .and_modify(|entries| {
+                let key = (
+                    get_relative_placement(first_qargs, second_qargs),
+                    (key1.clone(), key2.clone()),
+                );
+                entries.insert(key, is_commuting);
+                self.current_cache_entries += 1;
+            })
+            .or_insert_with(|| {
+                let mut entries = HashMap::with_capacity(1);
+                let key = (
+                    get_relative_placement(first_qargs, second_qargs),
+                    (key1, key2),
+                );
+                entries.insert(key, is_commuting);
+                self.current_cache_entries += 1;
+                entries
+            });
+        Ok(is_commuting)
+    }
+
+    #[allow(clippy::too_many_arguments)]
+    fn commute_matmul(
+        &self,
+        py: Python,
+        first_op: &OperationRef,
+        first_params: &[Param],
+        first_qargs: &[Qubit],
+        second_op: &OperationRef,
+        second_params: &[Param],
+        second_qargs: &[Qubit],
+    ) -> PyResult<bool> {
+        // Compute relative positioning of qargs of the second gate to the first gate.
+        // Since the qargs come out the same BitData, we already know there are no accidential
+        // bit-duplications, but this code additionally maps first_qargs to [0..n] and then
+        // computes second_qargs relative to that. For example, it performs the mappings
+        //  (first_qargs, second_qargs) = ( [1, 2], [0, 2] ) --> ( [0, 1], [2, 1] )
+        //  (first_qargs, second_qargs) = ( [1, 2, 0], [0, 3, 4] ) --> ( [0, 1, 2], [2, 3, 4] )
+        // This re-shuffling is done to compute the correct kronecker product later.
+        let mut qarg: HashMap<&Qubit, Qubit> = HashMap::from_iter(
+            first_qargs
+                .iter()
+                .enumerate()
+                .map(|(i, q)| (q, Qubit(i as u32))),
+        );
+        let mut num_qubits = first_qargs.len() as u32;
+        for q in second_qargs {
+            if !qarg.contains_key(q) {
+                qarg.insert(q, Qubit(num_qubits));
+                num_qubits += 1;
+            }
+        }
+
+        let first_qarg: Vec<Qubit> = Vec::from_iter((0..first_qargs.len() as u32).map(Qubit));
+        let second_qarg: Vec<Qubit> = second_qargs.iter().map(|q| *qarg.get(q).unwrap()).collect();
```

Comment:
Just use regular indexing here, since it will panic anyway if the Qarg doesn't exist.
```suggestion
        let second_qarg: Vec<Qubit> = second_qargs.iter().map(|q| qarg[q]).collect();
```

Reply:
Done in 12d0a05